packages/aws-cdk-lib/aws-batch/lib/managed-compute-environment.ts

```diff
@@ -1131,15 +1141,61 @@ export class FargateComputeEnvironment extends ManagedComputeEnvironmentBase imp
   }
 }
 
-function renderInstances(types?: ec2.InstanceType[], classes?: ec2.InstanceClass[], useOptimalInstanceClasses?: boolean): string[] {
-  const instances = [];
+function checkArchitectureConsistency(
+  newInstanceType: ec2.InstanceType,
+  existingTypes: ec2.InstanceType[],
+  existingClasses: ec2.InstanceClass[],
+): void {
+  const newArchitecture = newInstanceType.architecture;
+
+  for (const existingType of existingTypes ?? []) {
+    if (existingType.architecture !== newArchitecture) {
+      throw new Error(`Cannot add instance type ${newInstanceType.toString()} with architecture ${newArchitecture}. It conflicts with existing instance type ${existingType.toString()} with architecture ${existingType.architecture}.`);
+    }
+  }
+
+  for (const existingClass of existingClasses ?? []) {
+    const tempInstanceType = new ec2.InstanceType(`${existingClass.toString()}.large`);
+    const existingClassArchitecture = tempInstanceType.architecture;
+    if (existingClassArchitecture !== newArchitecture) {
+      throw new Error(`Cannot add instance type ${newInstanceType.toString()} with architecture ${newArchitecture}. It conflicts with existing instance class ${existingClass} with architecture ${existingClassArchitecture}.`);
+    }
+  }
+}
 
+function renderInstances(
+  scope: Construct, types?: ec2.InstanceType[], classes?: ec2.InstanceClass[], useOptimalInstanceClasses?: boolean): string[] {
+  const instances = [];
+  let hasArmInstances = false;
+  let hasX86Instances = false;
   for (const instanceType of types ?? []) {
     instances.push(instanceType.toString());
+    if (instanceType.architecture === ec2.InstanceArchitecture.ARM_64) {
+      hasArmInstances = true;
+    } else {
+      hasX86Instances = true;
+    }
   }
+
   for (const instanceClass of classes ?? []) {
     instances.push(instanceClass);
+    const tempInstanceType = new ec2.InstanceType(`${instanceClass.toString()}.large`);
+    if (tempInstanceType.architecture === ec2.InstanceArchitecture.ARM_64) {
+      hasArmInstances = true;
+    } else {
+      hasX86Instances = true;
+    }
   }
+
+  if (hasArmInstances && hasX86Instances) {
+    throw new Error('Cannot mix ARM and x86 instance types or classes');
```

Comment:
let's fallback to the warning here as well instead of error

Reply:
If we don't throw the error here, there will be an error when deploying which is the error shown in the original issue. Just making sure, would we only want a warning even though there will be an error deploying?

Reply:
yeah, i think i also was leaning towards throwing but since it affects both EC2 and EKS environment, want to be cautious in not causing regression with unexpected failures for existing customers but still keeping the validation to let them know about the expected behaviour beforehand, to confirm i did check the documentation , we can add it above the validation for future reference. 
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-batch-computeenvironment-computeresources.html#cfn-batch-computeenvironment-computeresources-instancetypes

Reply:
Updated here: [016a89a](https://github.com/aws/aws-cdk/pull/31510/commits/016a89a18ec4efba9f8d2d2d2a2f219419a913d1)!